src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -930,12 +935,15 @@ internal ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, Cancella
                 ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
             }
 
-            // Allocate whatever the pool gives us so we can write, this also marks the
-            // state as writing
-            AllocateWriteHeadIfNeeded(0);
+            CompletionData completionData;
+            ValueTask<FlushResult> result;
 
             lock (_sync)
             {
+                // Allocate whatever the pool gives us so we can write, this also marks the
+                // state as writing
+                AllocateWriteHeadIfNeeded(0);
```

Comment:
You moved `AllocateWriteHeadIfNeeded` under the lock here, but there are other usages that are not under the lock.

Reply:
True, the pattern is different for WriteAsync vs the other places we call this. WriteAsync is doing a bunch of operations, so trying to prevent disposal during that. GetMemory and GetSpan return data to the caller and there’s no good way to lock that operation, if the caller tries to Complete during the call to those, it’s undefined behavior. It’s worth looking at if we should revisit throwing on Complete but the user experience sucks

Reply:
Are you trying to protect against the writer being disposed?

Reply:
Why allow it at all?

Reply:
I want to preserve the behavior we had before making this change to overrwrite WriteAsync. Dispose would wait if there was writing in progress before (because it takes the lock). Later we can decide if we need to throw if there's writing happening when dispose is called.

Reply:
Is this part of the behaviour worth preserving? It relied on undefined behaviour as you stated above and is against Pipe usage rules.

Reply:
For this PR yes, for 2 reasons:
- It'll break tests in ASP.NET Core (that need to be fixed anyways) and tests here that are harder to fix because we share a pipe field.
- I don't like the behavior of the Complete throwing, the more I think about it, the other places should throw instead (write, flush etc) but it's not always possible (like if you wrote and left data in the buffer than completed).